determineversion blog post draft. (#1473)
google/osv.dev: gcp/appengine/blog/content/posts/using-the-determineversion-api.md

---
title: "Using the determineversion API to find C/C++ vulnerabilities"
date: 2023-07-18T19:16:34Z
draft: true
author: OSV Team
---
With the increasing incidence of software supply chain attacks, it is more important than ever for developers to understand the known vulnerabilities in their open source dependencies, regardless of the ecosystem of origin. Within the C/C++ ecosystem it is difficult to match dependencies to vulnerabilities for a few of reasons:

- C/C++ does not have a centralized package manager like npm or pyPI
- Software projects typically pull in C/C++ by submodules or vendoring
- Source code identifiers (e.g. git hashes) are the best way to identify libraries, but vulnerabilities are typically associated to versions, not git hashes

> OSV has had C/C++ vulnerability data from OSS-Fuzz keyed on git hashes from day 1. However, a remaining challenge for C/C++ users is being able to accurately identify the closest upstream git hash of their C/C++ dependencies in order to make use of this vulnerability data. The OSV team is committed to bridging the gap between what C/C++ users need and the constraints of the ecosystem and today we are announcing a tool that can help–the determineversion API. 

Reviewer: The lede is a little buried -- the announcement is kind of lost at the end of the paragraph. I'd move it to the beginning of the paragraph.

Author: Good point. I have moved some text to the first line.

<!--more-->

## What is the determineversion API?
The [determineversion API](https://google.github.io/osv.dev/post-v1-determineversion/) is a new, experimental API endpoint for OSV’s API. The goal of the determineversion API is to help users determine the likely version of their vendored C/C++ dependencies. Once dependency versions are known, our other API endpoints can return reliable vulnerability information. 

> We are actively seeking feedback on the determineversion API. If you have a project that you would like to scan for vulnerabilities, please follow along with this walkthrough and let us [know about your experience](https://github.com/google/osv.dev/issues/new). Your feedback can help us improve the vulnerability matching experience for other C/C++ developers. 

Reviewer: This paragraph seems out of place here. Would it work better at the end?

Author: Moved.

## What projects can use the determineversion API?
The set of C/C++ repositories currently supported by the determineversion API are the ones being fuzzed by [OSS-Fuzz](https://github.com/google/oss-fuzz). This is because OSV currently only contains commit-level vulnerability information for C/C++ projects via OSS-Fuzz. If a project has been fuzzed by OSS-Fuzz, you’ll be able to use the tool, but you should understand that there may be additional vulnerabilities in your dependencies that weren’t found by or reported through OSS-Fuzz. You can check a project against the API's current limitations by following the steps in our [documentation](https://google.github.io/osv.dev/post-v1-determineversion/#available-libraries). 

The OSV team is also working on expanding C/C++ coverage by including commit-level details for vulnerabilities from CVEs in the National Vulnerability Database(NVD). Once this is complete, our vulnerability matching for C/C++ will be comprehensive. 

## Let's try the determineversion API
To try the determineversion API, you will first need the following:

> 1. A local copy of the [osv.dev repository](https://github.com/google/osv.dev). This includes a tool that will simplify the use of the API.

Reviewer: Minor nit, but these don't need to be numbered; bullets would be better

> 1. A local copy of the [osv.dev repository](https://github.com/google/osv.dev). This includes a tool that will simplify the use of the API.

Author: ```suggestion
- A local copy of the [osv.dev repository](https://github.com/google/osv.dev). This includes a tool that will simplify the use of the API.
```

> 2. A local copy of the C/C++ project or projects for which you want to determine the version(s). This local copy can be vendored, submoduled, or you can clone a copy of our example project, [libxml2](https://github.com/GNOME/libxml2).

Author: ```suggestion
- A local copy of the C/C++ project or projects for which you want to determine the version(s). This local copy can be vendored, submoduled, or you can clone a copy of our example project, [libxml2](https://github.com/GNOME/libxml2).
```

                        "severity": "HIGH"
                    },
                    "database_specific": {
                        "source": "https://github.com/google/oss-fuzz-vulns/blob/main/vulns/libxml2/OSV-2021-777.yaml"
                    }
                    }
                ],
                "schema_version": "1.4.0"
                }
            ]
            }
            ```
5. Finally, we consider the response and draw conclusions. 
    
    To be sure we have caught any potential vulnerabilities, we could make further queries for other versions with similar scores. It is our opinion that this is generally unnecessary, but it could be done. 

    In this case, even if the actual version is not `2.11.3`, we can be fairly confident that the vulnerability that we found ([OSV-2021-777](https://osv.dev/vulnerability/OSV-2021-777)) is in our local copy of libxml2. This is because there is overlap between the other likely versions of libxml2 and the versions vulnerable to OSV-2021-777.

    By running our tool and making one additional API call, we now are fairly confident that my local version of libxml2 has known vulnerability OSV-2021-777.

> ## Try for yourself

Author: ```suggestion

## Try for yourself
```

---
title: "Using the determineversion API to find C/C++ vulnerabilities"
date: 2023-07-18T19:16:34Z
draft: true
author: OSV Team
---

With the increasing incidence of software supply chain attacks, it is more important than ever for developers to understand the known vulnerabilities in their open source dependencies, regardless of the ecosystem of origin. The determineversion API is OSV's newest tool that will help C/C++ developers match their dependencies to known vulnerabilities. 

> Within the C/C++ ecosystem it is difficult to match dependencies to vulnerabilities for a few of reasons:

Author: ```suggestion
Within the C/C++ ecosystem it is difficult to match dependencies to vulnerabilities for a few reasons:
```

                        "v2.9.12",
                        "v2.9.13",
                        "v2.9.14",
                        "v2.10.0",
                        "v2.10.1",
                        "v2.10.2",
                        "v2.10.3",
                        "v2.10.4",
                        "v2.11.0",
                        "v2.11.1",
                        "v2.11.2",
                        "v2.11.3",
                        "v2.11.4"
                    ],
                    "ecosystem_specific": {
                        "severity": "HIGH"
                    },
                    "database_specific": {
                        "source": "https://github.com/google/oss-fuzz-vulns/blob/main/vulns/libxml2/OSV-2021-777.yaml"
                    }
                    }
                ],
                "schema_version": "1.4.0"
                }
            ]
            }
            ```
5. Finally, we consider the response and draw conclusions. 
    
    To be sure we have caught any potential vulnerabilities, we could make further queries for other versions with similar scores. It is our opinion that this is generally unnecessary, but it could be done. 

>     In this case, even if the actual version is not `2.11.3`, we can be fairly confident that the vulnerability that we found  ([OSV-2021-777](https://osv.dev/vulnerability/OSV-2021-777)) is in our local copy of libxml2. This is because there is overlap between the other likely versions of libxml2 and the versions vulnerable to OSV-2021-777.

Author: ```suggestion
    In this case, even if the actual version is not `2.11.3`, we can be fairly confident that the vulnerability that we found ([OSV-2021-777](https://osv.dev/vulnerability/OSV-2021-777)) is in our local copy of libxml2. This is because there is overlap between the other likely versions of libxml2 and the versions vulnerable to OSV-2021-777.
```

>     In this case, even if the actual version is not `2.11.3`, we can be fairly confident that the vulnerability that we found in our sample API call ([OSV-2021-777](https://osv.dev/vulnerability/OSV-2021-777)) is in our local copy of libxml2. This is because there is overlap between the other likely versions of libxml2 and the versions vulnerable to OSV-2021-777.

Author: ```suggestion
    In this case, even if the actual version is not `2.11.3`, we can be fairly confident that the vulnerability that we found  ([OSV-2021-777](https://osv.dev/vulnerability/OSV-2021-777)) is in our local copy of libxml2. This is because there is overlap between the other likely versions of libxml2 and the versions vulnerable to OSV-2021-777.
```

                        "severity": "HIGH"
                    },
                    "database_specific": {
                        "source": "https://github.com/google/oss-fuzz-vulns/blob/main/vulns/libxml2/OSV-2021-777.yaml"
                    }
                    }
                ],
                "schema_version": "1.4.0"
                }
            ]
            }
            ```
5. Finally, we consider the response and draw conclusions. 
    
    To be sure we have caught any potential vulnerabilities, we could make further queries for other versions with similar scores. It is our opinion that this is generally unnecessary, but it could be done. 

    In this case, even if the actual version is not `2.11.3`, we can be fairly confident that the vulnerability that we found in our sample API call ([OSV-2021-777](https://osv.dev/vulnerability/OSV-2021-777)) is in our local copy of libxml2. This is because there is overlap between the other likely versions of libxml2 and the versions vulnerable to OSV-2021-777.

    By running our tool and making one additional API call, we now are fairly confident that my local version of libxml2 has known vulnerability OSV-2021-777.

> 

Author: ```suggestion
```

---
title: "Using the determineversion API to find C/C++ vulnerabilities"
date: 2023-07-18T19:16:34Z
draft: true
author: OSV Team
---
With the increasing incidence of software supply chain attacks, it is more important than ever for developers to understand the known vulnerabilities in their open source dependencies, regardless of the ecosystem of origin. Within the C/C++ ecosystem it is difficult to match dependencies to vulnerabilities for a few of reasons:

- C/C++ does not have a centralized package manager like npm or pyPI
- Software projects typically pull in C/C++ by submodules or vendoring
- Source code identifiers (e.g. git hashes) are the best way to identify libraries, but vulnerabilities are typically associated to versions, not git hashes

> OSV has had C/C++ vulnerability data from OSS-Fuzz keyed on git hashes from day 1. However, a remaining challenge for C/C++ users is being able to accurately identify the closest upstream git hash of their C/C++ dependencies in order to make use of this vulnerability data.The OSV team is committed to bridging the gap between what C/C++ users need and the constraints of the ecosystem and today we are announcing a tool that can help–the determineversion API. 

Reviewer 2: ```suggestion
OSV has had C/C++ vulnerability data from OSS-Fuzz keyed on git hashes from day 1. However, a remaining challenge for C/C++ users is being able to accurately identify the closest upstream git hash of their C/C++ dependencies in order to make use of this vulnerability data. The OSV team is committed to bridging the gap between what C/C++ users need and the constraints of the ecosystem and today we are announcing a tool that can help–the determineversion API. 
```

<!--more-->

## What is the determineversion API?
The [determineversion API](https://google.github.io/osv.dev/post-v1-determineversion/) is a new, experimental API endpoint for OSV’s API. The goal of the determineversion API is to help users determine the likely version of their vendored C/C++ dependencies. Once dependency versions are known, our other API endpoints can return reliable vulnerability information. 

## What projects can use the determineversion API?
The set of C/C++ repositories currently supported by the determineversion API are the ones being fuzzed in OSS-Fuzz. This is because OSV currently only contains commit-level vulnerability information for C/C++ projects via OSS-Fuzz. If a project has been fuzzed by OSS-Fuzz, you’ll be able to use the tool, but you should understand that there may be additional vulnerabilities in your dependencies that weren’t found by or reported through OSS-Fuzz. You can check a project’s eligibility by following the steps in our [docs](https://google.github.io/osv.dev/post-v1-determineversion/#available-libraries). 

> The OSV team is also working on expanding C/C++ coverage by attaching commit-level details to the NVC/CVE database.Once released, we are hopeful that many of these gaps will no longer exist. Even small security steps can make a positive difference, and we hope that the current limitations will not prevent you from trying the new API endpoint. 

Reviewer 3: I'd reword this more around that the NVD/CVE support will offer an extremely comprehensive solution to C/C++ vulnerability management. 

The current one is not a "small step" -- OSS-fuzz is already fuzzing most of the top C/C++ packages out there. So I'd remove the last sentence in this paragraph completely.

Author: This is a fair point. I have rephrased.

                        "v2.9.11",
                        "v2.9.12",
                        "v2.9.13",
                        "v2.9.14",
                        "v2.10.0",
                        "v2.10.1",
                        "v2.10.2",
                        "v2.10.3",
                        "v2.10.4",
                        "v2.11.0",
                        "v2.11.1",
                        "v2.11.2",
                        "v2.11.3",
                        "v2.11.4"
                    ],
                    "ecosystem_specific": {
                        "severity": "HIGH"
                    },
                    "database_specific": {
                        "source": "https://github.com/google/oss-fuzz-vulns/blob/main/vulns/libxml2/OSV-2021-777.yaml"
                    }
                    }
                ],
                "schema_version": "1.4.0"
                }
            ]
            }
            ```
5. Finally, we consider the response and draw conclusions. 

Our copy of libxml2 returned a number of versions with similar confidence scores. 
- Version 2.11.3, 0.718
- Version 2.11.4, 0.718
- Version 2.11.2, 0.707
- Version 2.11.0, 0.697

> To be sure we have caught any potential vulnerabilities, we could make further queries for these other versions and see if any other vulnerabilities are found. 

Reviewer 2: I would add a note here saying that it is very likely the version/s that have the highest confidence score is very likely the closest version to the actual code, more than the small difference in the score indicates. E.g. even though there's only a 0.011 difference in score between 2.11.2 and 2.11.4, it is overwhelmingly more likely (95%+) that 2.11.4 is going to be the closer version to the actual code)

To put it another way, the score is more of how similar the submitted code is to the version, and less a probability of what version the submitted code is. So while it's safer to also make further queries against the other versions, most of the time it's fine to just rely on the top result that's returned. 

Not sure how to word this in a concise and understandable way.

Additionally, we can be fairly confident that the vulnerability that we found in our sample API call ([OSV-2021-777](https://osv.dev/vulnerability/OSV-2021-777)) is in our local copy of libxml2. This is because there is overlap between the other likely versions of libxml2 and the versions vulnerable to OSV-2021-777.

> By running one Go module and making one additional API call, we now are fairly confident that my local version of libxml2 has known vulnerability OSV-2021-777.

Reviewer 2: Is this go module referring to the `indexer-api-caller`?, the term "go module" sounds a bit weird here.

I would just have this say "by making two API calls, first to determineversion, and then to query" or something similar.

If you would like to keep the mention of the tool, then something like: "by running out helper tool and making one additional API call"

Author: Yes, that was what I was referring to.

                    ],
                    "ecosystem_specific": {
                        "severity": "HIGH"
                    },
                    "database_specific": {
                        "source": "https://github.com/google/oss-fuzz-vulns/blob/main/vulns/libxml2/OSV-2021-777.yaml"
                    }
                    }
                ],
                "schema_version": "1.4.0"
                }
            ]
            }
            ```
5. Finally, we consider the response and draw conclusions. 

Our copy of libxml2 returned a number of versions with similar confidence scores. 
- Version 2.11.3, 0.718
- Version 2.11.4, 0.718
- Version 2.11.2, 0.707
- Version 2.11.0, 0.697

To be sure we have caught any potential vulnerabilities, we could make further queries for these other versions and see if any other vulnerabilities are found. 

Additionally, we can be fairly confident that the vulnerability that we found in our sample API call ([OSV-2021-777](https://osv.dev/vulnerability/OSV-2021-777)) is in our local copy of libxml2. This is because there is overlap between the other likely versions of libxml2 and the versions vulnerable to OSV-2021-777.

> By running one Go module and making one additional API call, we now are fairly confident that my local version of libxml2 has known vulnerability OSV-2021-777.

Author: ```suggestion
By running our tool and making one additional API call, we now are fairly confident that my local version of libxml2 has known vulnerability OSV-2021-777.
```
